x/incentive/keeper/rewards.go

```diff
@@ -86,3 +87,42 @@ func (k *Keeper) synchronizeClaim(
 
 	return claim
 }
+
+// GetSynchronizedClaim fetches a claim from the store and syncs rewards for all
+// rewarded sourceIDs.
+func (k Keeper) GetSynchronizedClaim(
+	ctx sdk.Context,
+	claimType types.ClaimType,
+	owner sdk.AccAddress,
+) (types.Claim, bool) {
+	claim, found := k.GetClaim(ctx, claimType, owner)
+	if !found {
+		return types.Claim{}, false
+	}
+
+	// Fetch all source IDs from indexes
+
+	var sourceIDs []string
+	k.IterateRewardIndexesByClaimType(ctx, claimType, func(rewardIndexes types.TypedRewardIndexes) bool {
+		sourceIDs = append(sourceIDs, rewardIndexes.CollateralType)
+		return false
+	})
+
+	adapter := k.GetSourceAdapter(claimType)
+	accShares := adapter.OwnerSharesBySource(ctx, owner, sourceIDs)
+
+	var sortedSourceIDs []string
+	for sourceID := range accShares {
+		sortedSourceIDs = append(sortedSourceIDs, sourceID)
+	}
+
+	// Sort source IDs to ensure deterministic order of claim syncs
+	sort.Strings(sortedSourceIDs)
+
+	// Synchronize claim for each source ID
+	for _, sourceID := range sortedSourceIDs {
+		claim = k.synchronizeClaim(ctx, claim, sourceID, owner, accShares[sourceID])
+	}
```

Comment:
suggestion: I think it could be marginally safer to iterate through `sourceIDs` rather than the sorted keys. If an adapter has a bug and doesn't return all provided sourceIDs, this could silently skip a sync. Whereas if `sourceIDs` was iterated through and the map didn't contain the key this could panic and make the issues visible.